### ISSUE TITLE
Import browser bookmarks

### ISSUE THREAD

**Issue by Treora** [WebMemex/webmemex-extension#29]
Get all of the browser bookmarks and add them to the database, while marking them as being imported.

Each bookmark would map to a page in the database. Perhaps we should also try to pull the visits to these pages from the history, even if one does not choose to import the full history.

How to represent that a page is bookmarked is a remaining question. Adding a flag to the page itself sounds easy but is perhaps the wrong place. Adding a new type of item that links to the page is probably a more logical solution, to keep the pages themselves separated from the user's organisation.

The folder structure, tags and descriptions of bookmarks are important, but perhaps we will deal with those only later on.

**Comment by blackforestboi** [comment]
Do I understand it correctly?

There would be 3 types of objects:

1) Visits 
2) Bookmarks
3) Content

Visits and bookmarks are containing metainformation dependend on the object type (i.e. `date_added` (bookmark) or `VisitTime` (History Item), whereas the Content is having all the data like text, title, keywords, tags etc)?

From what I understand, this structure seems to be a good one when wanting to fetch a list of history items or bookmarks, without having to load all the content of those elements in it. And if desired so, the full content can still be loaded afterwards. correct?
This would also make it easy to do the full-text search optional, as we could leave minimal information inside the visit and bookmark item, so that they is better searchable just by title & URL.

Would a content object still have all the references to the visit/bookmark objects, so filtering by time or bookmarks is possible?

**Comment by wimpykid26** [comment]
Hello @Treora @oliversauter , I would like to work on this if you let me. From what I understood :
1. Create a small frontend component in 'Overview' that will listen to the event 'importBookmarks'
2. A background script that will execute.
3. Import bookmarks using the Webextension Bookmarks API.
4. Convert the Array into pageDocs and visitDocs.
5. Map them into model and then push them into the db. 

In case of visitDocs, we'll have to map the bookmark to the corresponding history entry and then obtain the number of times It has been visited.
Since the inclusion of 'Content' has popped up, I am having trouble understanding about how this new model will go into the db as it currently has pageDocs and visitDocs as properties.
Am I missing something?. Thank you in advance :)

**Comment by blackforestboi** [comment]
@wimpykid26 

Maybe it helps to look into how the history is imported and stored: https://github.com/WebMemex/webmemex-extension/commit/5ef86e7a050d3285e3353fbf3737b4b22b6a5a44

I just want to give this as a pointer to understand the general mechanics of how the download works, and would advise waiting for @Treora's input from a technical perspective before starting to develop. 
This is because I don't know in which format the bookmark is eventually stored - as a new object or as information appended to the pageDocs item.

**Comment by Treora** [comment]
Hi @wimpykid26, thanks for helping! Your approach sounds good. I would just remove one step and add one step:

> 1. Create a small frontend component in 'Overview' that will listen to the event 'importBookmarks'

Not exactly sure what you meant here. Perhaps you meant to make a component that _triggers_ the `importBookmarks` script? Anyhow, the idea was to not put this in the overview but in the options pages, and this idea is described in #17. For the scope of this issue, it would be fine to just create the script that performs the import process as you described, and exposes this. Like Oliver said, have a look at the history import as an example.

The step to add would be to create a new type of document in the model that represents the bookmark itself. Besides adding the page ('pageDoc') to the database, we would like to somehow remember that this page is more important to the user than the others they visited.

So we would roughly get the three types of docs as Oliver listed them, except that his term 'Content' is called a page doc, and a page doc does not necessarily contain the content. The browser does not keep any content of the bookmarked page except its title, so this title is all we can store in the page doc, just like when importing history. If we want to have the full content of the page in the database, we can try to fetch it again from the web, but that is another issue (#28).

As for Oliver's last question (sorry I never answered this @oliversauter):
> Would a content object still have all the references to the visit/bookmark objects, so filtering by time or bookmarks is possible?

In principle the references would point from a visit/bookmark doc to the page doc, rather than the other way around. To search for pages with certain content, that are bookmarked or visited in a particular time, we could create indexes that combine the knowledge of the different types of documents. Some things still have to be figured out still here, as this may not doable out-of-the-box with PouchDB, so we may end up combining multiple queries, adopting a graph database, creating other indexing solutions, or resorting to adding those reverse references to the page docs after all.

**Comment by wimpykid26** [comment]
@Treora how do I trigger my 'import bookmarks' script to check if it is working properly?.

**Comment by Treora** [comment]
You could of course just make a simple button in any page while testing it. Myself, I would probably just make the function a global variable during development: `window.testImportBookmarks = ....`, and then call it from the debugger console; which you want to look at anyway to check error messages.

**Comment by Treora** [comment]
By the way, I think [this](https://github.com/WorldBrain/Research-Engine/blob/3d32a35a42ad47f57c01161404af1ec8a8a6c835/src/js/import_bookmarks.js#L12) is a smart trick I noticed in the WorldBrain code:
> `chrome.bookmarks.getRecent(100000 , function(bookmarks) {`

No need to walk down trees of bookmarks this way. It does not import the bookmark folder structure though, but we have no way to represent folders yet anyhow so we could just start with this.

**Comment by blackforestboi** [comment]
@wimpykid26 What is the progress of this issue?

**Comment by wimpykid26** [comment]
@oliversauter commented on the PR, waiting for @Treora to give his feedback and suggest how I should go about making the functions generic and differentiate between history and bookmark items.

**Comment by Treora** [comment]
This issue is still waiting for (my) attention, along with several others.. It would be good to tackle a few import-from-browser-related things together (bookmark import (#29), its UI #17, filter by bookmarks (#39), maybe keeping in sync with browser bookmarks (#12), fetching&analysing the documents (#28)), but I came to realise these may not be high enough on my priority queue soon; there is too much to do, and it seems better to try focus on implementing essentials and core concepts first. I am considering to mark these features as 'for later'.

In case somebody is willing and capable of taking the lead on tackling these interrelated issues, that could also be an option. It may however also technically be a wiser decision to first get the essentials right and stabilise data models before adding these browser integrations.

CC @oliversauter: this choice of prioritisation seems most relevant to you. You probably still desire most of these features quite strongly for use in WorldBrain, I guess?

**Comment by blackforestboi** [comment]
> CC @oliversauter: this choice of prioritisation seems most relevant to you. You probably still desire most of these features quite strongly for use in WorldBrain, I guess?

Yes indeed I intend to get this done soon. @poltak already started exploring this. He will probably redo the whole import bookmarks from scratch, but could need some support in understanding how you thought about the current data model. 

Would be good if you guys talk about your intended data model anyway, so we can collaboratively document it and that there are 2 persons understanding it entirely.

**Comment by Treora** [comment]
@poltak: thanks for thinking along ([over here](https://github.com/WebMemex/webmemex-extension/pull/47#issuecomment-296068677))!

> If my understanding is correct, after reading through #29, the main difference here should be that we'll want to save some kind of bookmark doc instead of a visit doc. Is my understanding correct?
> So:
> - `import-history` module: grab everything from history, generate **visit** + page docs
> - `import-bookmarks` module: grab everything from bookmarks, generate **bookmark** + page docs

Correct.
Small note: I also considered to have the import-bookmarks module call the `import-history` module to import the visits to the bookmarked pages. Conceptually, importing bookmarks would then be more like: "import my history, but only the bookmarked pages". This would make a better fit with the timeline view, which shows one's visits.

> If so, let's decide some structure for bookmark docs?

Yes.

Note that the question about doc structure depends on the question how we conceptually look at bookmarks: does one bookmark a URL as such (the location), whatever is served there, or does one bookmark the page that was there at time of bookmarking? Much of the thinking about our data model boils down to how we work around what is perhaps the most annoying design problem of the web, namely that we refer to documents by their (possibly temporary) location.

It is imortant to realise here that bookmark docs are not only there to represent the bookmarks from the browser. People should be able to create new bookmarks, which may store more data than the ones created by browsers. A bookmark should be able to refer to either a specific (version of a) page (and possibly a specific part of a page), or to a URL without further specificity. Bookmarks imported from the browser will have to be unspecific, while bookmarks created by the webmemex can be made more specific.

For a view on the future direction: I would like to keep the option open to use the Web Annotation data model for bookmarks, see [this part of the spec](https://www.w3.org/TR/2017/REC-annotation-model-20170223/#motivation-and-purpose). It would nicely allow increasing specificity, e.g. bookmarking a sentence in a page. I think our current data model could still differ, but it is nice if it is designed to allow for easy migration to this standard later on (i.e. the conceptual model matches).

> Bookmark doc structure could possibly look like:
> 
> ```
> {
>   _id, 
>   url, // URL of the bookmarked page 
>   bookmarkStart, // timestamp of when the site was bookmarked
>   importedFromBrowserBookmarks, // timestamp of when the import happened
>   page: { _id }, // ID of the page doc which this bookmark is associated with
> }
> ```

- Visits (will) have a `visitEnd`, but bookmarks probably not, so `bookmarkStart` could just be called `createdAt` or something. It could in theory even be left out if we have the creation timestamp also stored in the `id`, but I'm happy to keep it there explicitly, as I am not so fond of relying on this timestamp in the id (they're just there for the pragmatic reason of not needing a secondary index in the database for the timestamp).
- `importedFromBrowserBookmarks` is of course an optional field, just an extra tag to distinguish those that were imported.

> An additional unrelated question: When will the `url` in a visit doc and the `url` in that visit's associated page doc differ? All the URLs in the associated docs i'm looking at in my local pouchDB seems to be the same.

Quite a related question actually. Currently the urls in visit and page won't differ. It did not seem obvious enough to me that they have to be the same however. I was thinking ahead of a possible situation where a page has multiple locations (which could happen through e.g. some page deduplication strategy), and the visit retrieved the page through one of them. Remember URLs are used both to point at locations, and to identify a page. My idea was to rather save too much than too little, and keep the model flexible.

The same thinking may apply to bookmarks: The bookmark may refer to a page, of which we know just the URL and the title. Or it refers to a location. Or both, kind of. So when importing bookmarks, perhaps we should store the URL in both the bookmark and its implied page again?

For clarification: in theory, we could even consider not creating any page doc at all when importing bookmarks, as we know little about it since the existence of a page is merely implied by the bookmark. I think the reason for creating the page doc is to encode exactly that little information we have: there was supposedly a page at this URL with this title at this time (saving the title is useful anyway, and saving the title in the bookmark itself seems wrong).

**Comment by poltak** [comment]
@Treora thanks for all the in-depth responses across these threads.

> I also considered to have the import-bookmarks module call the import-history module to import the visits to the bookmarked pages. Conceptually, importing bookmarks would then be more like: "import my history, but only the bookmarked pages".

What about bookmarks that don't appear in history? Say a user visits a page on their phone, bookmarks it, then it gets synced to their desktop browser's bookmarks, and they run import bookmarks later? 

> does one bookmark a URL as such (the location), whatever is served there, or does one bookmark the page that was there at time of bookmarking?

If we choose to consider bookmark data at the time they were bookmarked and want to get page data for that time (via the timestamp), we would have to do this via some web page archiving service, right? 

But I think both of these use-cases are valid across users. A user might bookmark the frontpage of a news site, fully aware that when they visit that bookmark in the future the content will have changed. However, the same user might bookmark a single article, expecting any time that they return the article contains the same content as before (although regardless of the user's expectation, the article author may make revisions, or completely remove it). (You touched on this stuff in the [dedup PR](https://github.com/WebMemex/webmemex-extension/pull/60#issuecomment-285871701) actually)

Yes we can add extra data to our model to take into account the user's expectation with the bookmarks they make within the extension, but for importing existing browser bookmarks I think it's maybe easier (for now at least) just to use what the given URL points to at the time of import?

**Comment by Treora** [comment]
> What about bookmarks that don't appear in history? Say a user visits a page on their phone, bookmarks it, then it gets synced to their desktop browser's bookmarks, and they run import bookmarks later?

Good point. The simple concept that search just applies a filter to your browsing history does not fit anymore then. We could perhaps add the act of bookmarking as an event in the browsing history to make it fit though.

In longer term it may be worth thinking what would be an intuitive way to present the search, conceptually. Now we present it as a timeline of things you visited, but the concept of the project as a whole is rather about owning your own piece of web. Showing that web in a timeline is often helpful as it provides context that other representations (e.g. the usual ranked result list) lack, but we may encounter more cases where the concepts don't match up completely. A discussion for another time though.

> If we choose to consider bookmark data at the time they were bookmarked and want to get page data for that time (via the timestamp), we would have to do this via some web page archiving service, right?

Yes. Filling in that page stub from web archives sounds reasonable, would be a fun side-project. I have no plans for that myself however.

> I think it's maybe easier (for now at least) just to use what the given URL points to at the time of import?

You mean to use it as the content of the page object that the bookmark points to? Yes, slightly easier, but I think it should hardly be more effort to follow the 'time aware' approach and just create a new page object for the fetched document. The framework is already in place: you would just treat a fetched bookmark like any other logged visit, except that it happens in the background. You could then still choose an aggressive deduplication strategy to make them one object again, if desired.

**Comment by poltak** [comment]
@Treora Do you already have some idea of how you plan to use the Web Annotation Data Model with bookmarks docs? From trying to understand that spec page, we would need to restructure all the docs to be something like the examples given and have them differ on `motivation` type? So for bookmarks, something like:
```javascript
{
  "_id": "bookmark/:timestamp/:nonce",
  "type": "Annotation",
  "motivation": "bookmarking",
  "body": [
    {
      "type": "TextualBody", // Extracted text
      "value": "....",
      "purpose": "describing" // ??
    },
    {
      "type": "??", // Extracted meta-data
      "value": "...",
      "purpose": "describing" // ??
    },
    {
      "type": "??", // Favicon
      "value": "...", // Data URI
      "purpose": "describing" 
    },
    {
      "type": "??", // Page title
      "value": "Example title",
      "purpose": "describing" // ??
    }
  ],
  "target": "http://example.com" // Same as the current URL key
}
```

There doesn't seem to be any `motivation` types appropriate for `page` or `visit` docs, or the types to use for page title, extracted metadata, and all the other additional information we are storing now in page docs (they seem to fit into `body`...). 

I'm very confused about the plans with this web annotation data model, so this example may be/probably is completely wrong...

**Comment by Treora** [comment]
@poltak: I was thinking of something much simpler, no big structural
changes. The bookmarking 'annotation' would not have a body, which I
think the spec also suggests. Anyhow, feel free to forget about the spec
for now, we can rename properties to follow it later.

For comparison, a quick (probably incorrect) sketch of what I think a
bookmark object could roughly look like:

    {
        target: {
            source: "https://xkcd.com/123",
            state: {
                // a TimeState with bookmarking date
                cached: "uri-for-our-saved-page-if-any", // URI would in
    some way correspond to the page _id.
            }
        },
        motivation: "bookmarking"
    }

Not sure if this makes most sense though. We could stick extra custom
properties into it if we like, e.g to store the pageId directly.

**Comment by Treora** [comment]
Closing this issue; features like this one have been fading off the roadmap for a while. Maybe some day..
